handlers/hierarchy.go

```diff
@@ -306,35 +306,37 @@ type flatNodes struct {
 	defaultOrder map[string]int
 }
 
-func (n flatNodes) addWithoutChildren(val hierarchy.Child, i int) {
-	if !val.HasData {
+func (n *flatNodes) addWithoutChildren(val hierarchy.Child) {
+	if n == nil || !val.HasData {
 		return
 	}
 
-	n.list[i] = &hierarchy.Child{
+	n.list = append(n.list, &hierarchy.Child{
 		Label:   val.Label,
 		Links:   val.Links,
 		HasData: val.HasData,
 		Order:   val.Order,
-	}
+	})
 }
 
-func (n flatNodes) addWithChildren(val hierarchy.Child, i int) {
-	if len(n.list) < i {
+func (n *flatNodes) addWithChildren(val hierarchy.Child) {
+	if n == nil {
 		return
 	}
-
-	n.list[i] = &hierarchy.Child{
+	n.list = append(n.list, &hierarchy.Child{
 		Label:            val.Label,
 		Links:            val.Links,
 		HasData:          val.HasData,
 		Order:            val.Order,
 		NumberofChildren: val.NumberofChildren,
-	}
+	})
 }
 
 // hasOrder returns true if and only if all child items in the list have a non-nil order values
-func (n flatNodes) hasOrder() bool {
+func (n *flatNodes) hasOrder() bool {
+	if n == nil || n.list == nil {
```

Comment:
I think it's pretty unusual to be checking a pointer receiver for a nil value. Not saying change it necessarily as it won't hurt as such, but if this is an issue that actually comes up it suggest the surrounding code is bad. It's not really a common pattern you see, it's the sort of thing that would be idiomatic to do always if it was.

Reply:
It's not something that was required, I just started checking everything that I saw. The change that was actually needed is to check that n.list contains non-nil items (which should be fixed anyway by the changes to append values instead of writing them in particular indices).
I can remove the check for the nil pointer receiver, as you say if it is nil, there's something really wrong.

Reply:
Personally I would remove them but it's completely up to you.